handler/src/main/java/io/netty/handler/ssl/SniHandler.java

```diff
@@ -321,6 +321,19 @@ private void replaceHandler(ChannelHandlerContext ctx, Selection selection) {
         }
     }
 
+    /**
+     * Users may override this method to implement custom behavior. The default
+     * implementation will simply replace {@code this} {@link SniHandler} instance
+     * with the given {@link SslHandler} using the provided {@link SSLEngine}.
+     *
+     * BEWARE: The {@code hostname} argument can be {@code null}.
+     */
+    protected void replaceHandler(ChannelHandlerContext ctx, String hostname,
+            SslContext sslContext, SSLEngine sslEngine) throws Exception {
+
+        ctx.pipeline().replace(this, SslHandler.class.getName(), SslContext.newHandler(sslEngine));
```

Comment:
I agree, the responsibility is very murky. Your suggestion works but I noticed that SslHandler has a bunch of setter methods and I'm wondering if the current `private replaceHandler(...)` method isn't "wrong"?

I mean... If I don't use SNI then I'm fiddling with the SslHandler directly. Why not give the user the same option with SNI?

How about this... Change the `#select()` method to call a protected method like this...

```java
protected void calledFromSelect(ChannelHandlerContext ctx, 
    String hostname, SslContext sslContext) {

  replaceHandler(ctx, new Selection(hostname, sslContext));
}

private void replaceHandler(ChannelHandlerContext ctx, Selection selection) {
  // existing logic
}
```

If the user overrides `calledFromSelect(...)` they're on their own and have full control over what they plan to do.